aptos-move/aptos-aggregator/src/delta_change_set.rs

```diff
@@ -76,27 +76,66 @@ impl DeltaOp {
         }
     }
 
-    /// Aggregates another delta into `self`.
-    pub fn merge_with(&mut self, other: DeltaOp) -> PartialVMResult<()> {
+    /// Applies self on top of previous delta, merging them together. Note
+    /// that the strict ordering here is crucial for catching overflows
+    /// correctly.
+    pub fn merge_onto(&mut self, previous_delta: DeltaOp) -> PartialVMResult<()> {
         use DeltaUpdate::*;
 
-        Ok(match (self.update, other.update) {
-            (Plus(value), Plus(other_value)) => {
-                let new_value = addition(value, other_value, self.limit)?;
-                self.update = Plus(new_value);
-            }
-            (Plus(value), Minus(other_value)) | (Minus(other_value), Plus(value)) => {
-                if value >= other_value {
-                    let new_value = subtraction(value, other_value)?;
-                    self.update = Plus(new_value);
-                } else {
-                    let new_value = subtraction(other_value, value)?;
-                    self.update = Minus(new_value);
+        Ok(match previous_delta.update {
+            Plus(value) => {
+                // Before we proceed to merging deltas, make sure we verify that
+                // `self` delta did not overflow when shifted by +value. At the
+                // same time, we can compute new history values for `self` delta.
+                let other_max_positive = addition(value, self.max_positive, self.limit)?;
+                let other_min_negative = subtraction(self.min_negative, value).or(Ok(0))?;
+
+                // If check has passed, update the value.
+                match self.update {
+                    Plus(other_value) => {
+                        let new_value = addition(value, other_value, self.limit)?;
```

Comment:
Same question here and below whenever an error is possible

Reply:
here we shortcutIf error, `?` propagates it to the caller